stdlib/parray.mc

```diff
@@ -0,0 +1,369 @@
+-- of O(log n).
+
+-- Pure Array
+type PA a
+con PAData  : all a. (Int,a,a,a,a,a,a,a,a,a,a) -> PA a
+con PANode  : all a. (Int,Int,PA a,PA a,PA a,PA a,PA a,PA a,PA a,PA a,PA a,PA a) -> PA a
+con PAEmpty : all a. () -> PA a
+con PANext  : all a. (PA a) -> PA a
+
+let empty = PAEmpty()
+
+-- Helper function for prettyPA, for generating a pretty print string for
+-- debugging
+recursive
+let workPrettyPA = lam ident. lam pa. lam f.
+  match pa with PAData(i,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9) then
+   join ["\n", ident, "PAData(", int2string i,",",
+           f x0, ",", f x1, ",", f x2, ",", f x3, ",",
+           f x4, ",", f x5, ",", f x6, ",", f x7, ",",
+           f x8, ",", f x9, ")"] else
+  match pa with PANode(i,l,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9) then
+   let g = lam pa. workPrettyPA (concat ident "  ") pa f in
+   join ["\n", ident, "PANode(", int2string i, ",", int2string l, ",",
+                    g x0, ",", g x1, ",", g x2, ",", g x3, ",",
+                    g x4, ",", g x5, ",", g x6, ",", g x7, ",",
+                    g x8, ",", g x9, ")"] else
+  match pa with PAEmpty() then "E" else
+  match pa with PANext(pa2) in workPrettyPA ident pa2 f
+end
+
+-- `prettyPA pa f` is a simple pretty printer, mostly for internal debugging,
+-- where f : a -> String is a pretty printer for the element of the array.
+-- The function returns a string.
+let prettyPA = workPrettyPA ""
+
+recursive
+let workAddPA = lam pa. lam y. lam l.
+  match pa with PAEmpty() then
+    if leqi l 1
+      then PAData(1,y,y,y,y,y,y,y,y,y,y)
+      else PANode(0,l,workAddPA empty y (divi l 10),
+           empty,empty,empty,empty,empty,empty,empty,empty,empty)
+  else
+  match pa with PAData(1,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(2,x0,y ,x2,x3,x4,x5,x6,x7,x8,x9) else
+  match pa with PAData(2,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(3,x0,x1,y ,x3,x4,x5,x6,x7,x8,x9) else
+  match pa with PAData(3,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(4,x0,x1,x2,y ,x4,x5,x6,x7,x8,x9) else
+  match pa with PAData(4,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(5,x0,x1,x2,x3,y ,x5,x6,x7,x8,x9) else
+  match pa with PAData(5,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(6,x0,x1,x2,x3,x4,y ,x6,x7,x8,x9) else
+  match pa with PAData(6,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(7,x0,x1,x2,x3,x4,x5,y ,x7,x8,x9) else
+  match pa with PAData(7,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(8,x0,x1,x2,x3,x4,x5,x6,y ,x8,x9) else
+  match pa with PAData(8,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9)
+           then PAData(9,x0,x1,x2,x3,x4,x5,x6,x7,y ,x9) else
+  match pa with PAData(9,x0,x1,x2,x3,x4,x5,x6,x7,x8,x9) then
+           if eqi l 0 then PANode(1,10,PAData(
+                   10,x0,x1,x2,x3,x4,x5,x6,x7,x8,y),
+                   empty,empty,empty,empty,empty,empty,empty,empty,empty)
+           else
+            PANext(PAData(10,x0,x1,x2,x3,x4,x5,x6,x7,x8,y)) else
```

Comment:
I think these could be written less verbosely, as
```
match pa with PAData t in
switch t.0
case 1 then PAData{t with #label"0" = 2, #label"2" = y}
case 2 then PAData{t with #label"0" = 3, #label"3" = y}
...
end
```
Other parts in the file could also benefit from the `switch` sugar

Reply:
Well, that is a matter of taste. I will think about it.

Reply:
I have fixed this now when I think it makes it less verbose, and still clear. Thanks for the feedback!